components/service/sync-logins/src/main/java/mozilla/components/service/sync/logins/DefaultLoginValidationDelegate.kt

```diff
@@ -24,27 +26,105 @@ class DefaultLoginValidationDelegate(
     private val scope: CoroutineScope = CoroutineScope(IO)
 ) : LoginValidationDelegate {
 
-    @Suppress("ComplexMethod") // This method is not actually complex
-    override fun validateCanPersist(login: Login): Deferred<Result> {
+    /**
+     * Uses saved [Login]s and new [Login] to return a [Result] for knowing whether an existing
+     * login record can and should be updated or created.
+     */
+    @Suppress("ComplexMethod")
+    override fun getNewLoginPersistenceAsync(newLogin: Login): Deferred<Result> {
         return scope.async {
-            try {
-                // We're setting guid=null here in order to trigger de-duping logic properly.
-                // Internally, the library ensures to not dupe records against themselves
-                // (via a `guid <> :guid` check), which means we need to "pretend" this is a new record,
-                // in order for `ensureValid` to actually throw `DUPLICATE_LOGIN`.
-                storage.value.ensureValid(login.copy(guid = null))
-                Result.CanBeCreated
+            val savedLoginByGuid = newLogin.guid?.let { storage.value.get(it) }
+            val savedLoginByElements = storage.value.getStoredLoginByImportantFields(newLogin)
+            // Check the merged login with the new password to display correct UI error states
+            val loginToCheck =
+                (savedLoginByGuid ?: savedLoginByElements)?.mergeWithLogin(newLogin)
+                    ?.copy(password = newLogin.password) ?: newLogin
+            val result = try {
+                storage.value.ensureValid(loginToCheck)
+                when {
+                    // GUIDs don't match, but everything else important matches so we can merge and update
+                    savedLoginByElements != null -> {
+                        Result.CanBeUpdated
+                    }
+                    newLogin.guid.isNullOrEmpty() || savedLoginByGuid == null -> {
+                        Result.CanBeCreated
+                    }
+                    newLogin.guid != savedLoginByGuid.guid -> {
```

Comment:
If `storage.get` is functioning correctly, this is impossible. And I really hope `get` is functioning correctly :)